compiler/rustc_lint/src/empty_iterator.rs

```diff
@@ -0,0 +1,76 @@
+use crate::{EarlyContext, EarlyLintPass, LintContext};
+
+use rustc_ast::{Expr, ExprKind, Lit, LitKind};
+use rustc_errors::fluent;
+
+declare_lint! {
+    /// The `empty_for_iterator` lint checks for empty `for` iterator.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// for i in 10..0 { /* ... */ }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// It is usually a mistake to have a statement that has no effect.
+    pub EMPTY_FOR_ITERATOR,
+    Warn,
+    "empty for iterator"
+}
+
+declare_lint_pass!(EmptyForIterator => [EMPTY_FOR_ITERATOR]);
+
+impl EarlyLintPass for EmptyForIterator {
```

Comment:
Does it make sense to use late pass here instead, so, for example, contants can be evaluated (at least I think they can...)?

Reply:
Yes, @obeis please use late pass. Generally late pass should be used if you don't need anything specific to the AST (so always use late pass if you can implement it both as early or late)

Reply:
I think i can not use late pass because i need `rustc_ast::Expr` and `rust_ast::ExprKind`. Range expressions and for loops are desugared at HIR level.
see this [discuss](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/range.20expr.20include.20Path.20kind)